### PR TITLE
Introduce standard devicetree Kconfig symbols for drivers

### PR DESCRIPTION
**Pull request by galak**
The intent of this PR is to standardize and hopefully ease some of the boilerplate around driver Kconfig and devicetree.

We already have a number of drivers that utilize the following pattern:

```
DT_COMPAT_NXP_IMX_CSI := nxp,imx-csi

config VIDEO_MCUX_CSI
   bool "NXP MCUX CMOS Sensor Interface (CSI) driver"
   default $(dt_compat_enabled,$(DT_COMPAT_NXP_IMX_CSI))
   ...
```

The idea is to introduce drivers/Kconfig.dts that a script can generate by scanning the source tree for all dts bindings and build a list of `compatible` values from.  The above example than would look like:

```
DT_COMPAT_NXP_IMX_CSI := nxp,imx-csi

config DTS_HAS_NXP_IMX_CSI
        def_bool $(dt_has_compat,$(DT_COMPAT_NXP_IMX_CSI))

config DTS_HAS_NXP_IMX_CSI_ENABLED
        depends on DTS_HAS_NXP_IMX_CSI
        def_bool $(dt_compat_enabled,$(DT_COMPAT_NXP_IMX_CSI))
```

and the `drivers/video/Kconfig.mcux_csi` looks like:

```
config VIDEO_MCUX_CSI
	bool "NXP MCUX CMOS Sensor Interface (CSI) driver"
	default DTS_HAS_NXP_IMX_CSI_ENABLED
	depends on DTS_HAS_NXP_IMX_CSI
	...  
```

* We introduce `DTS_HAS_<sanitized compat>_ENABLED` to make it a little easier on driver writers and to catch errors like: `default $(dt_compat_enabled,${DT_COMPAT_FOOBAR})` (use of `{`,`}`  is incorrect.
* We introduce `DT_HAS_<sanitized compat>` so that only drivers that exist in a board's devicetree are exposed (for those drivers that are devicetree based).

Additionally by introducing standardize and generated Kconfig symbols for devicetree compatibles we create an association between devicetree/Kconfig/CMake that we can utilize in the future for various purposes like doc generation or testing.

### CODE REVIEW

**Comment by galak** [comment]
NOTE: This PR is based on https://github.com/zephyrproject-rtos/zephyr/pull/47158 to show the benefit for something like a build_all test in which just adding having the devicetree node added to the test gets the driver enabled and building.

**Comment by carlescufi** [comment]
API meeting:

- @carlescufi this will add lots of needless Kconfig options to document
- @gmarull mentions that this will not be compatible with OoT drivers
- @erwango mentions that it might be a bit burdensome: add the compat to the list, regenerate the Kconfig.dts file, etc.
- The list file will likely have conflicts often
- Caspar: Could this be done at build time based on your own compatibles? @nashif does not like the idea
- @nashif alternative: Use a Kconfig template but without autogenerating anything, just reducing the boilerplate

**Comment by galak** [comment]
> * @carlescufi this will add lots of needless Kconfig options to document

It doesn't look like `def_bool`s show in the Kconfig generated docs today?
 
> * @gmarull mentions that this will not be compatible with OoT drivers

Why OoT drivers do or don't in this area shouldn't impact things.  The drivers are OoT.  
 
> * @erwango mentions that it might be a bit burdensome: add the compat to the list, regenerate the Kconfig.dts file, etc.
> 
> * The list file will likely have conflicts often

1. The compat list would need to exist, one script will scan bindings dir and generate Kconfig.dts
2. Wonder how often we'll have conflicts if the file is sorted by name.  Likely new changes will not be next to each other.

> * Caspar: Could this be done at build time based on your own compatibles? @nashif does not like the idea
> 
> * @nashif alternative: Use a Kconfig template but without autogenerating anything, just reducing the boilerplate

**Comment by henrikbrixandersen** [comment]
> Why OoT drivers do or don't in this area shouldn't impact things. The drivers are OoT.

But OOT drivers need to be able to use the same infrastructure as used for in-tree drivers.

**Comment by gmarull** [comment]
In general I like the idea of standardizing these options, but my main concern is the autogenerated file. Ideas/comments:

- OOT drivers need to be able to use the same Kconfig patterns
- Offline file means more maintenance (e.g. extra CI checks, potential merge conflicts, etc.)
- I guess generating such file at runtime takes too much time?
- We could generate the file at runtime based on parsed DTS (smaller I guess) and handle non-existing options that have the `DT_...` pattern as `n` in Kconfig scripts.

**Comment by erwango** [comment]
On my side, I'm affraid this might be cryptic to new comers: 
How to guess that a script should be run after a new binding is added?

**Comment by galak** [comment]
I've updated this to now be handled at runtime.  So this should work for OoT and not require any manual intervention.

@gmarull, @henrikbrixandersen, @erwango take a look.

File: scripts/dts/gen_driver_kconfig_dts.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import argparse
+import os
+import sys
+import re
+from collections import defaultdict
+
+import yaml
+try:
+    # Use the C LibYAML parser if available, rather than the Python parser.
+    # This makes e.g. gen_defines.py more than twice as fast.
```
Comment (gmarull) [inline scripts/dts/gen_driver_kconfig_dts.py:16]: this message doesn't seem relevant here?

File: scripts/dts/gen_driver_kconfig_dts.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    # Returns a list with the paths to all bindings (.yaml files) in
+    # 'bindings_dirs'
+
+    binding_paths = []
+
+    for bindings_dir in bindings_dirs:
+        for root, _, filenames in os.walk(bindings_dir):
+            for filename in filenames:
+                if filename.endswith(".yaml") or filename.endswith(".yml"):
+                    binding_paths.append(os.path.join(root, filename))
```
Comment (gmarull) [inline scripts/dts/gen_driver_kconfig_dts.py:34]: suggestion: pathlib glob could likely simplify this (I think)

File: cmake/modules/dts.cmake [hunk 3]
```diff
@@ -233,6 +238,18 @@ if(SUPPORTS_DTS)
...
   endif()
 
+
+  execute_process(
+    COMMAND ${PYTHON_EXECUTABLE} ${GEN_DRIVER_KCONFIG_SCRIPT}
+    --kconfig-out ${DTS_KCONFIG}
+    --bindings-dirs ${DTS_ROOT_BINDINGS}
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+    RESULT_VARIABLE ret
+    )
```
Comment (gmarull) [inline cmake/modules/dts.cmake:248]: Maybe add a status message? Configure step is already cluttered enough, though...

File: cmake/modules/dts.cmake [hunk 4]
```diff
@@ -233,6 +238,18 @@ if(SUPPORTS_DTS)
     message(STATUS "Generated device_extern.h: ${DEVICE_EXTERN_H}")
   endif()
 
+
+  execute_process(
+    COMMAND ${PYTHON_EXECUTABLE} ${GEN_DRIVER_KCONFIG_SCRIPT}
```
Comment (gmarull) [inline cmake/modules/dts.cmake:243]: question: we are forcing to re-run configure stage if we add a new compatible, is it like this now?

**Review by gmarull — COMMENTED** [review]
LGTM, a few questions/suggestions. I've tested the changes and I didn't notice a build time increase.

Btw, I've tested OOT case, and it seems to be working well, which is great:
https://github.com/zephyrproject-rtos/example-application/pull/20

File: scripts/dts/gen_driver_kconfig_dts.py (hunk 2 above)
Comment (galak) [reply scripts/dts/gen_driver_kconfig_dts.py:34]: This is pretty much lifted from edtlib.py so going to leave it alone for now.

File: cmake/modules/dts.cmake (hunk 3 above)
Comment (galak) [reply cmake/modules/dts.cmake:248]: Yeah, don't know what the status message would say/add and how much anyone pays any attention to them.

File: scripts/dts/gen_driver_kconfig_dts.py (hunk 1 above)
Comment (galak) [reply scripts/dts/gen_driver_kconfig_dts.py:16]: will fix, copy/pasted from gen_defines.py ;)

**Review by mbolivar-nordic — COMMENTED** [review]
I think this is a great idea. Have you given any thought to how you want to document it?

**Comment by galak** [comment]
> I think this is a great idea. Have you given any thought to how you want to document it?

What aspect of docs?   I haven't yet given that much thought.

**Comment by galak** [comment]
One question I have is related to keeping a "cached" copy of the in `dts/Kconfig.dts`. 

Positives:
* reduces execution time of script
* reduces how big the generated file is, so a bit easier to read
* Allows for "editor" completion of Kconfig symbols.
* Able to see some level of what changed, got added between updates

Negatives:
* requires update from time to time (maybe just before or after a release) of `dts/Kconfig.dts`

**Comment by galak** [comment]
@henrikbrixandersen, @carlescufi , @MaureenHelm , @nashif, @erwango, @gmarull, @mbolivar-nordic what else needs to be done here?

I know @mbolivar-nordic mentioned docs, so looking for peoples input on what doc updates we should make for this.

Anything else?

**Comment by galak** [comment]
> I can't find anywhere to put this documentation, which makes me sad. One more bit of proof that our driver documentation is missing a lot of real-world details.

Agreed, was thinking how we don't have a "how to write a driver" guide.

**Comment by galak** [comment]
> @henrikbrixandersen, @carlescufi , @MaureenHelm , @nashif, @erwango, @gmarull, @mbolivar-nordic what else needs to be done here?

Updated to just have the `DT_<COMPAT>_ENABLED` symbol.  Its all we actually need.  It doesn't make sense to allow a driver to be compiled in if at least one instance isn't enabled.  Otherwise we'll just get compile errors (and this addresses an issue with singleton devices that use `DEVICE_DT_DEFINE`.

**Comment by microbuilder** [comment]
> > I can't find anywhere to put this documentation, which makes me sad. One more bit of proof that our driver documentation is missing a lot of real-world details.
> 
> Agreed, was thinking how we don't have a "how to write a driver" guide.

I wouldn’t mind working on that if you and @mbolivar-nordic are OK to collaborate on the DT side.

File: drivers/video/Kconfig.mcux_csi [hunk 5]
```diff
@@ -3,9 +3,8 @@
 # Copyright (c) 2019, Linaro Limited
 # SPDX-License-Identifier: Apache-2.0
 
-DT_COMPAT_NXP_IMX_CSI := nxp,imx-csi
-
 config VIDEO_MCUX_CSI
 	bool "NXP MCUX CMOS Sensor Interface (CSI) driver"
-	default $(dt_compat_enabled,$(DT_COMPAT_NXP_IMX_CSI))
+	default DTS_HAS_NXP_IMX_CSI_ENABLED
```
Comment (henrikbrixandersen) [inline drivers/video/Kconfig.mcux_csi:8]: How come `DTS_` and not `DT_` like we use in other devicetree APIs?

**Review by henrikbrixandersen — COMMENTED** [review]
Looks really good, one small question:

File: drivers/video/Kconfig.mcux_csi (hunk 5 above)
Comment (galak) [reply drivers/video/Kconfig.mcux_csi]: No real reason, I changed it to `DT_`

**Review by henrikbrixandersen — DISMISSED** [review]
Looks good, thank you.

**Review by microbuilder — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]
LGTM

File: drivers/video/Kconfig.mt9m114 [hunk 6]
```diff
@@ -6,5 +6,7 @@
 config VIDEO_MT9M114
 	bool "MT9M114 Aptina CMOS digital image sensor"
 	depends on I2C
+	depends on DT_HAS_APTINA_MT9M114_ENABLED
+	default DT_HAS_APTINA_MT9M114_ENABLED
```
Comment (pdgendt) [inline drivers/video/Kconfig.mt9m114:10]: What is the difference with or advantage over:
```suggestion
	default y
```

Comment (galak) [reply drivers/video/Kconfig.mt9m114:10]: @pdgendt there isn't one.  It made a bit more sense before I added the `depends on`.  But with that, it should just be `default y`.

**Review by pdgendt — DISMISSED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks for cleaning this up @galak!

**Review by pdgendt — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Pull request merged**